// @allmaps/cli/dist/commands/iiif/image-ids.js
import { Command } from '@commander-js/extra-typings';
import { fromZodError } from 'zod-validation-error';
import { parseJsonInput, printString } from '../../lib/io.js';
import { parseIiif } from '../../lib/iiif.js';
import { addParseIiifOptions } from '../../lib/options.js';
function isCollection(item) {
    return (item.type === 'collection' && item.embedded === false && 'items' in item);
}
function collectImageIds(parsedIiifs) {
    const imageIds = [];
    for (const parsedIiif of parsedIiifs) {
        if (parsedIiif.type === 'image') {
            imageIds.push(parsedIiif.uri);
        }
        else if (parsedIiif.type === 'manifest') {
            imageIds.push(...parsedIiif.canvases.map((canvas) => canvas.image.uri));
        }
        else if (parsedIiif.type === 'collection' && 'items' in parsedIiif) {
            const collectionImageIds = parsedIiif.items
                .filter((item) => isCollection(item))
                .map((item) => collectImageIds([item]));
            imageIds.push(...collectionImageIds.flat());
        }
    }
    return imageIds;
}
export function imageIds() {
    const command = addParseIiifOptions(new Command('image-ids')
        .argument('[files...]')
        .summary('read all IIIF Image IDs from IIIF resources')
        .description('Reads all IIIF Image IDs from IIIF resources'));
    return command.action(async (files, options) => {
        const jsonValues = await parseJsonInput(files);
        const parsedIiifs = [];
        for (const jsonValue of jsonValues) {
            try {
                const parseIiifOptions = {
                    fetchCollections: options.fetchCollections,
                    fetchManifests: options.fetchManifests,
                    fetchImages: options.fetchImages,
                    fetchAll: options.fetchAll,
                    maxDepth: options.fetchMaxDepth
                };
                const parsedIiif = await parseIiif(jsonValue, parseIiifOptions);
                parsedIiifs.push(parsedIiif);
            }
            catch (err) {
                if (err instanceof Error && err.name === 'ZodError') {
                    const zodError = err;
                    const validationError = fromZodError(zodError);
                    console.error(validationError.toString());
                }
                else if (err instanceof Error) {
                    console.error(err.message);
                }
            }
        }
        const imageIds = new Set(collectImageIds(parsedIiifs));
        printString([...imageIds].join('\n'));
    });
}
