// @allmaps/cli/dist/lib/iiif.js
import { IIIF } from '@allmaps/iiif-parser';
const defaultParseOptions = {
    fetchCollections: false,
    fetchManifests: false,
    fetchImages: false,
    fetchAll: false,
    maxDepth: Number.POSITIVE_INFINITY
};
function fetchFn(input, init) {
    console.warn('Fetching', input);
    return fetch(input, init);
}
export async function parseIiif(sourceIiif, options) {
    if (options && options.fetchAll) {
        options = {
            ...options,
            fetchCollections: true,
            fetchManifests: true,
            fetchImages: true
        };
    }
    else {
        options = {
            ...defaultParseOptions,
            ...options
        };
    }
    const parsedIiif = IIIF.parse(sourceIiif);
    if (parsedIiif.type === 'collection') {
        await parsedIiif.fetchAllItems({ ...options, fetchFn });
    }
    else if (parsedIiif.type === 'manifest' && options && options.fetchImages) {
        await parsedIiif.fetchAllItems(fetchFn);
    }
    return parsedIiif;
}
function generateImageService(image) {
    const profile = image.supportsAnyRegionAndSize ? 'level1' : 'level0';
    const type = `ImageService${image.majorVersion}`;
    return {
        id: image.uri,
        profile,
        type
    };
}
export function generateManifest(id, images) {
    return {
        '@context': 'http://iiif.io/api/presentation/3/context.json',
        id,
        type: 'Manifest',
        items: images.map((image, index) => ({
            id: `${id}/canvases/${index}`,
            type: 'Canvas',
            width: image.width,
            height: image.height,
            items: [
                {
                    type: 'AnnotationPage',
                    id: `${id}/canvases/${index}/annotation-page`,
                    items: [
                        {
                            type: 'Annotation',
                            id: `${id}/canvases/${index}/annotation`,
                            motivation: 'painting',
                            target: `${id}/canvases/${index}`,
                            body: {
                                type: 'Image',
                                id: image.uri,
                                width: image.width,
                                height: image.height,
                                service: [generateImageService(image)]
                            }
                        }
                    ]
                }
            ]
        }))
    };
}
